### PR TITLE
Add Array.shuffle

### PR DESCRIPTION
**Pull request by dbuenzli**
Today I got tired of c&p this bit. 

```ocaml
val shuffle : rand:(int -> int) -> 'a array -> unit
(** [shuffle rand a] randomly permutes [a]'s elements using [rand]
    for randomness.

    [rand] must be such that a call to [rand n] returns a uniformly
    distributed random number in the range \[[0];[n-1]\].
    {!Random.int} can be used for this (do not forget to
    {{!Random.self_init}initialize} the generator).
```

Implemented via [Fisher-Yates](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm).

### CODE REVIEW

**Review by nojb — APPROVED** [review]
A useful addition (and non completely trivial to implement). LGTM.

A second approval from a core developer is needed.

**Review by dra27 — APPROVED** [review]
Agree with @nojb on both counts (indeed, I pasted it into a script last week, too...)

**Comment by dbuenzli** [comment]
I would say *it is* trivial to implement. However you are extremely likely to screw up if you roll your own algorithm.

Thanks for the approvals.

File: stdlib/array.ml
Comment (alainfrisch) [inline stdlib/array.ml]: Unless I miss something obvious,  `j` might not be a valid index. I see this is mandated by the interface, but this makes the function as unsafe as `unsafe_get/unsafe_set`. I think one should use checked accesses.

**Comment by alainfrisch** [comment]
Two questions:
  - Has it been considered to put the function in the `Random` module instead (with a variant using the global generator and another one using an explicit state)? This would also allow getting rid of the bound checks that will need to be added to make the function safe.
  - What about adding versions that operate on `bytes` and one 1d bigarrays?

File: stdlib/array.mli
Comment (alainfrisch) [inline stdlib/array.mli]: Something should be said about the function implementation a **uniform** distribution of the permutation (assuming the uniform distribution of `rand`).

Comment (alainfrisch) [reply stdlib/array.mli]: Is there a reason to use a different word for the function name and in its docstring (shuffle vs permute)?  I'm not against it, just checking that this is intentional.

File: stdlib/array.ml
Comment (nojb) [reply stdlib/array.ml]: Oops, good catch!

Comment (alainfrisch) [reply stdlib/array.ml]: Eagerly waiting for automated code review :-)

image

image

Comment (dbuenzli) [reply stdlib/array.ml]: Chat GPT style, always slightly trivially wrong.

File: stdlib/array.mli
Comment (dbuenzli) [reply stdlib/array.mli]: I added a sentence to the doc string regarding the uniform distribution. 

Regarding shuffle vs permute. I'm not attached to the name both seem to be in use. A quick grab bag, it seems containers, [`scala`](https://www.scala-lang.org/api/2.13.5/scala/util/Random$.html#shuffle[T,C](xs:IterableOnce[T])(implicitbf:scala.collection.BuildFrom[xs.type,T,C]):C) [`ruby`](https://apidock.com/ruby/v2_5_5/Array/shuffle),  [C++](https://en.cppreference.com/w/cpp/algorithm/random_shuffle), lots of other libraries in other language use `shuffle` and base uses `permute`. Sherlocode tells me [shuffle](https://sherlocode.com/?q=shuffle) is not uncommon in ocaml code and here are the results for [permute](https://sherlocode.com/?q=permute).

One (rather weak) argument in favor of `shuffle` is that `permute` could be taken as swapping two elements, though in a reading context I don't think it would be. I'd be more in favour of `permutation` if we were not doing it in place.

If people want to debate this to death, please do I will gladly take the name people compromise on.

**Comment by dbuenzli** [comment]
> * Has it been considered to put the function in the `Random` module instead (with a variant using the global generator and another one using an explicit state)?

I prefer if people can plug the random value generator they wish.

> * What about adding versions that operate on `bytes` and one 1d bigarrays?

I'm not deeply against it but given the churn and amount of cut and pasting such a simple addition entails, I'd rather have the correct name, doc string and implementation in first. I was already crying over that Float.Array business and the whole build system generation machinery. 

Note that none of these have the sorting functions either. If people are interested in making these interfaces more uniform I think it's better to do that in one go rather than have point releases each sprinkling a new function here and there.

File: stdlib/array.mli
Comment (smuenzel) [inline stdlib/array.mli:332]: We could make the `rand` argument optional, and default to `Random.int`

Comment (xavierleroy) [reply stdlib/array.mli:332]: This would cause a circular dependency between Array and Random.  But it would be fine if this function was in Random.

Comment (gasche) [reply stdlib/array.mli:332]: Personally I am fine with the current choice:
- I think that if this was in the Random module, it would be more natural to have `Random.shuffle : 'a array -> 'a array` and `Random.State.shuffle : Random.State.t -> 'a array -> 'a array`
- but I see benefits in abstracting over the random generator
- the convenience cost for the user is fairly small, and it forces them to wonder if they want to pass a specific state or rely on the implicit domain state

I see the abstraction over a sampling function as arguably fairly similar to abstracting over a comparison function for sorting functions.

Comment (dbuenzli) [reply stdlib/array.mli:332]: I prefer to have sources of effects spelled out explicitly; less questions asked on reading the code. 

Regarding having it in the `Random` module, as mentioned above, I prefer to let users plug their own if they wish so. Especially if one needs long term stability guarantees that the stdlib `Random` module does not provide.

**Comment by dbuenzli** [comment]
Are @alainfrisch and @smuenzel happy about the answers that were given ?

File: Changes
Comment (gasche) [inline Changes]: (and Alain I guess, he found the safety bug)

**Review by gasche — APPROVED** [review]

**Pull request merged**
